### PR TITLE
Fixed #3731 - replace inserted variables

### LINKED ISSUE #3731: 7.9.1 Email: Lead "Insert Variable" does not insert selected data when sending
Create an email template, 
insert variable "Account/Lead/Lead First name"
"$lead_first_name" appears in preview boxes
Select "Send Text Only" check box

When sending, the receivers email only displays "$lead_first_name"

#### Expected Behavior
Variable data should be inserted in email before sending

### PR DESCRIPTION
**Pull request by gymad**
## Description

	replace email variables
## How To Test This

please see in related issue #3731
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: modules/Emails/EmailsController.php [hunk 1]
```diff
@@ -154,6 +161,116 @@ public function action_send()
...
+    }
+
+    /**
+     * Request validation before replace bean variables,
+     * see log to check validation problems
+     *
+     * @param array $request
+     * @return bool
+     */
+    protected function isValidRequestForReplaceEmailVariables($request) {
```
Comment (daniel-samson) [inline modules/Emails/EmailsController.php]: PSR-2 Formating - move the { to the next line

File: modules/Emails/EmailsController.php [hunk 2]
```diff
@@ -154,6 +161,116 @@ public function action_send()
...
+        return $ok;
+    }
+
+    /**
+     * Add a message to log
+     *
+     * @param string $msg
+     * @param string $level
+     */
+    private function log($msg, $level = 'info') {
```
Comment (daniel-samson) [inline modules/Emails/EmailsController.php]: PSR-2 Formating - move the { to the next line

File: modules/OutboundEmailAccounts/OutboundEmailAccounts.php [hunk 3]
```diff
@@ -43,6 +43,11 @@
 require_once('modules/OutboundEmailAccounts/OutboundEmailAccounts_sugar.php');
 class OutboundEmailAccounts extends OutboundEmailAccounts_sugar {
```
Comment (daniel-samson) [inline modules/OutboundEmailAccounts/OutboundEmailAccounts.php]: PSR-2 Formating - move the { to the next line

File: modules/OutboundEmailAccounts/OutboundEmailAccounts.php [hunk 4]
```diff
@@ -43,6 +43,11 @@
...
 
+	/**
+	 * @var string
+	 */
+	public $mail_smtppass;
+
 	function __construct(){
```
Comment (daniel-samson) [inline modules/OutboundEmailAccounts/OutboundEmailAccounts.php]: PSR-2 Formating - move the { to the next line

File: modules/Campaigns/WizardNewsletter.php [hunk 5]
```diff
@@ -393,20 +393,31 @@
...
-        'description' => $prospectLst->description,
-        'type' => $prospectLst->list_type,
-        'count' => $prospectLst->get_entry_count(),
-    );
-    $targetListDataArray[] = $nxt;
-    $targetListDataAssoc[$prospectLst->id] = $nxt;
+if($targetList) {
+    $targetListData = array();
+    foreach ($targetList as $prospectLst) {
+        $nxt = array(
```
Comment (daniel-samson) [inline modules/Campaigns/WizardNewsletter.php]: what is nxt? can we use a more descriptive name here?

File: modules/Emails/EmailsController.php [hunk 6]
```diff
@@ -154,6 +161,116 @@ public function action_send()
...
+     * @param array $request
+     * @return Email
+     */
+    protected function replaceEmailVariables(Email $email, $request) {
+
+        // request validation before replace bean variables
+
+        if ($this->isValidRequestForReplaceEmailVariables($request)) {
+
+            $macro_nv = array();
```
Comment (daniel-samson) [inline modules/Emails/EmailsController.php]: what does nv mean? Can we make this a little more descriptive?

File: modules/Campaigns/WizardMarketing.php [hunk 7]
```diff
@@ -271,8 +271,14 @@
...
-foreach($outboundEmailAccounts = BeanFactory::getBean('OutboundEmailAccounts')->get_full_list() as $outboundEmailAccount) {
-    $outboundEmailLabels[$outboundEmailAccount->id] = $outboundEmailAccount->name;
+$outboundEmailLabels = array();
+$outboundEmailAccounts = BeanFactory::getBean('OutboundEmailAccounts')->get_full_list();
+if ($outboundEmailAccounts) {
+    foreach ($outboundEmailAccounts as $outboundEmailAccount) {
+        $outboundEmailLabels[$outboundEmailAccount->id] = $outboundEmailAccount->name;
+    }
+} else {
+    $GLOBALS['log']->warn('No any outbound email account');
```
Comment (daniel-samson) [inline modules/Campaigns/WizardMarketing.php]: Can we change the message to something like "There are no outbound email accounts available. Please ensure that the email settings are configured correctly"?

File: modules/Campaigns/WizardNewsletter.php [hunk 8]
```diff
@@ -393,20 +393,31 @@
...
+            //'type' => $prospectLst->type,
+            'description' => $prospectLst->description,
+            'type' => $prospectLst->list_type,
+            'count' => $prospectLst->get_entry_count(),
+        );
+        $targetListDataArray[] = $nxt;
+        $targetListDataAssoc[$prospectLst->id] = $nxt;
+    }
+} else {
+    $GLOBALS['log']->warn('No any target list for campaign');
```
Comment (daniel-samson) [inline modules/Campaigns/WizardNewsletter.php]: Can we change the message to something like "There are no outbound target lists available for campaign <name>."?

Or Perhaps "Unable to find a target list for campaign <name>.

File: modules/Campaigns/WizardNewsletter.php [hunk 9]
```diff
@@ -629,7 +640,7 @@ function validate_wiz_form(step){
 
 $ss->display(file_exists('custom/modules/Campaigns/tpls/WizardNewsletter.tpl') ? 'custom/modules/Campaigns/tpls/WizardNewsletter.tpl' : 'modules/Campaigns/tpls/WizardNewsletter.tpl');
 
-if(!$focus->id) {
+if(!$focus->id && isset($campaign_id) && $campaign_id) {
```
Comment (daniel-samson) [inline modules/Campaigns/WizardNewsletter.php]: Could we add a comment explaining why we need this code? For example What is check looking for and why do we need to unset this value?

File: modules/OutboundEmailAccounts/OutboundEmailAccounts.php [hunk 10]
```diff
@@ -51,7 +56,12 @@ public function save($check_notify = false) {
 		if(!$this->mail_smtppass && $this->id) {
 			$bean = new OutboundEmailAccounts();
 			$bean->retrieve($this->id);
-			$this->mail_smtppass = $bean->mail_smtppass;
+			if(!$bean->mail_smtppass) {
+				$GLOBALS['log']->warn("No mail SMTP password");
```
Comment (daniel-samson) [inline modules/OutboundEmailAccounts/OutboundEmailAccounts.php]: Can we change the message to something like "Unable to send email via SMTP using an empty password. Please ensure that the email settings are configured correctly"?

File: modules/Emails/EmailsController.php [hunk 11]
```diff
@@ -154,6 +161,116 @@ public function action_send()
...
+    /**
+     * Request validation before replace bean variables,
+     * see log to check validation problems
+     *
+     * @param array $request
+     * @return bool
+     */
+    protected function isValidRequestForReplaceEmailVariables($request) {
+
+        $ok = true;
```
Comment (daniel-samson) [inline modules/Emails/EmailsController.php]: OK can === false, which is a little confusing. Can we perhaps change this to $response?

File: modules/Emails/EmailsController.php (hunk 6 above)
Comment (gymad) [inline modules/Emails/EmailsController.php]: In fact I have no idea what does it mean, I just want to use consistent nameing for variables like:
modules/EmailMan.php:329,333,444,773,786,789,790,877
modules/EmailTemplate.php:416,493
test/test/modules/EmailTemplates/EmailTemplateTest.php:127,132

File: modules/Campaigns/WizardNewsletter.php (hunk 5 above)
Comment (gymad) [inline modules/Campaigns/WizardNewsletter.php]: I think "nxt" means "next" of the results, look the next line. btw it's only a guess, in my PR I just changed the indent of these line.

Comment (gymad) [inline modules/Campaigns/WizardNewsletter.php]: (next PR will contains the fix)

**Comment by Dillon-Brown** [comment]
Hi @binita1911, the files changed would be:

```
modules/Campaigns/WizardMarketing.php 
modules/Campaigns/WizardNewsletter.php
modules/Emails/Email.php
modules/Emails/EmailsController.php
modules/OutboundEmailAccounts/OutboundEmailAccounts.php
```

I can't guarantee that implementing the same fix that was done here will completely solve the issue though, there have been many fixes since 7.10.9.

**Pull request merged**